mono/mini/debugger-engine.c

```diff
@@ -612,6 +629,896 @@ mono_de_stop_single_stepping (void)
+	MonoDomain *domain;
+	MonoContext *ctx = rt_callbacks.tls_get_restore_state (tls);
+	MonoMethod *method;
+	SeqPoint sp;
+	MonoSeqPointInfo *info;
+	SingleStepReq *ss_req;
+
+	/* Skip the instruction causing the single step */
+	rt_callbacks.begin_single_step_processing (ctx, from_signal);
+
+	if (rt_callbacks.try_process_suspend (tls, ctx))
+		return;
+
+	/*
+	 * This can run concurrently with a clear_event_request () call, so needs locking/reference counts.
+	 */
+	ss_req = ss_req_acquire ();
+
+	if (!ss_req)
+		// FIXME: A suspend race
+		return;
+
+	if (mono_thread_internal_current () != ss_req->thread)
+		goto exit;
+
+	ip = (guint8 *)MONO_CONTEXT_GET_IP (ctx);
+
+	ji = get_top_method_ji (ip, &domain, (gpointer*)&ip);
+	g_assert (ji && !ji->is_trampoline);
+
+	if (log_level > 0) {
+		DEBUG_PRINTF (1, "[%p] Single step event (depth=%s) at %s (%p)[0x%x], sp %p, last sp %p\n", (gpointer) (gsize) mono_native_thread_id_get (), ss_depth_to_string (ss_req->depth), mono_method_full_name (jinfo_get_method (ji), TRUE), MONO_CONTEXT_GET_IP (ctx), (int)((guint8*)MONO_CONTEXT_GET_IP (ctx) - (guint8*)ji->code_start), MONO_CONTEXT_GET_SP (ctx), ss_req->last_sp);
+	}
+
+	method = jinfo_get_method (ji);
+	g_assert (method);
+
+	if (method->wrapper_type && method->wrapper_type != MONO_WRAPPER_DYNAMIC_METHOD)
+		goto exit;
+
+	/* 
+	 * FIXME:
+	 * Stopping in memset makes half-initialized vtypes visible.
+	 * Stopping in memcpy makes half-copied vtypes visible.
+	 */
+	if (method->klass == mono_defaults.string_class && (!strcmp (method->name, "memset") || strstr (method->name, "memcpy")))
+		goto exit;
+
+	/*
+	 * This could be in mono_de_ss_update method, but mono_find_next_seq_point_for_native_offset is pretty expensive method,
+	 * hence we prefer this check here.
+	 */
+	if (ss_req->user_assemblies) {
+		gboolean found = FALSE;
+		for (int k = 0; ss_req->user_assemblies[k]; k++)
+			if (ss_req->user_assemblies[k] == m_class_get_image (method->klass)->assembly) {
+				found = TRUE;
+				break;
+			}
+		if (!found)
+			goto exit;
+	}
+
+	/*
+	 * The ip points to the instruction causing the single step event, which is before
+	 * the offset recorded in the seq point map, so find the next seq point after ip.
+	 */
+	if (!mono_find_next_seq_point_for_native_offset (domain, method, (guint8*)ip - (guint8*)ji->code_start, &info, &sp)) {
+		g_assert_not_reached ();
+		goto exit;
+	}
+
+	il_offset = sp.il_offset;
+
+	if (!mono_de_ss_update (ss_req, ji, &sp, tls, ctx, method))
+		goto exit;
+
+	/* Start single stepping again from the current sequence point */
+	SingleStepArgs args = {
+		.method = method,
+		.ctx = ctx,
+		.tls = tls,
+		.step_to_catch = FALSE,
+		.sp = sp,
+		.info = info,
+		.frames = NULL,
+		.nframes = 0
+	};
+	mono_de_ss_start (ss_req, &args);
+
+	if ((ss_req->filter & STEP_FILTER_STATIC_CTOR) &&
+		(method->flags & METHOD_ATTRIBUTE_SPECIAL_NAME) &&
+		!strcmp (method->name, ".cctor"))
+		goto exit;
+
+	// FIXME: Has to lock earlier
```

Comment:
Have we put this PR through any desktop stress tests to prevent desktop regression? I’m pretty sure our locking behavior doesn’t change here but it’s difficult to statically tell with these many changes. I’m okay with merging without the testing, but this seem like something for QA to look at before next release.

Reply:
Yeah, prob should